File: fire-code-copilot/backend/app/sections.py
"""Section-number utilities shared by the retriever (amendment matching) and the citation
validator. Code sections are hierarchical dotted numbers (903, 903.2, 903.2.8, 903.2.8.4).
Treating them as plain strings is the root of two bugs we fix elsewhere:
  - "903.2" should match its child "903.2.8" for amendment precedence (ancestor governs);
  - "903.2" must NOT be considered "present" just because "903.2.8" is (substring false-positive).
These helpers give one normalization + a precise ancestor/descendant relation.
"""
from __future__ import annotations
import re

_KEYWORDS = re.compile(r"(?i)\b(sections?|sec\.?|tables?|chapters?|chap\.?|appendix|annex)\b")
_PURE_NUMBER = re.compile(r"^\d+[A-Z]?(?:\.\d+)*$")
_FIRST_NUMBER = re.compile(r"\d+(?:\.\d+)*")


def canonical(s: str | None) -> str:
    """Normalize a section reference so '§903.2.8', 'Section 903.2.8.', 'Table 903.2.8' and
    '903.2.8' all compare equal. Non-section labels (e.g. 'NFPA 13') are upper-cased as-is."""
    s = _KEYWORDS.sub(" ", s or "")
    s = s.replace("§", " ")
    s = re.sub(r"\s+", " ", s).strip().upper()
    return s.strip(" .,:;")


def numeric_tuple(s: str | None) -> tuple[str, ...] | None:
    """Return the dotted components of a pure section number, else None.

    '903.2.8' -> ('903','2','8'); 'NFPA 13' -> None (a standard label, not a dotted section)."""
    c = canonical(s)
    if not c or not _PURE_NUMBER.match(c):
        return None
    return tuple(c.split("."))


def relates(a: str | None, b: str | None) -> bool:
    """True if a and b are the same section, or one is an ancestor/descendant of the other
    (sharing at least chapter+section, so a bare chapter like '903' doesn't sweep in everything)."""
    ca, cb = canonical(a), canonical(b)
    if not ca or not cb:
        return False
    if ca == cb:
        return True
    ta, tb = numeric_tuple(ca), numeric_tuple(cb)
    if ta and tb:
        n = min(len(ta), len(tb))
        return n >= 2 and ta[:n] == tb[:n]
    return False


def section_tokens(text: str) -> set[str]:
    """Every distinct dotted section number that literally appears in `text` (canonicalized),
    e.g. from a chunk body. Used for word-boundary 'is this section present?' checks."""
    return {m.group(0) for m in _FIRST_NUMBER.finditer(text or "") if "." in m.group(0)}

File: fire-code-copilot/frontend/src/lib/sections.ts
/**
 * sections.ts — code-section string helpers (frontend mirror of the backend
 * `sections.relates`). Kept pure + dependency-free so it's trivial to test.
 */

/**
 * Canonicalize a section label for comparison: strip the "§" / "Section" /
 * "Table" prefixes, all whitespace, and any trailing dot, then uppercase.
 * e.g. "§ 903.2.8." -> "903.2.8", "Table 1020.1" -> "1020.1".
 */
function canon(s: string): string {
  return s
    .replace(/§/g, "")
    .replace(/\bSection\b/gi, "")
    .replace(/\bTable\b/gi, "")
    .replace(/\s+/g, "")
    .replace(/\.+$/, "")
    .toUpperCase();
}

/** True for a pure dotted-number section like "903", "903.2", "903.2.8". */
function isDotted(s: string): boolean {
  return /^\d+(\.\d+)*$/.test(s);
}

/**
 * Do two section labels refer to related provisions? True when they canonicalize
 * to the same value, OR when both are pure dotted numbers and one is an
 * ancestor/descendant of the other sharing at least two leading components
 * (e.g. "903.2.8" relates to "903.2.8.4", but "903.2" does not relate to "903.3").
 * Mirrors the backend `sections.relates`.
 */
export function sectionsRelate(a?: string, b?: string): boolean {
  if (!a || !b) return false;
  const ca = canon(a);
  const cb = canon(b);
  if (!ca || !cb) return false;
  if (ca === cb) return true;
  if (!isDotted(ca) || !isDotted(cb)) return false;

  const pa = ca.split(".");
  const pb = cb.split(".");
  const shorter = pa.length <= pb.length ? pa : pb;
  const longer = pa.length <= pb.length ? pb : pa;

  // The shorter must be a prefix of the longer, sharing >= 2 leading components.
  if (shorter.length < 2) return false;
  for (let i = 0; i < shorter.length; i++) {
    if (shorter[i] !== longer[i]) return false;
  }
  return true;
}
